runtime/service.go

```diff
@@ -769,6 +791,115 @@ func (s *service) GetVMMetadata(requestCtx context.Context, request *proto.GetVM
 	return &proto.GetVMMetadataResponse{Metadata: string(metadata)}, nil
 }
 
+func (s *service) createBalloonDevice(req *proto.CreateVMRequest) firecracker.BalloonDevice {
+	amountMB := req.BalloonDevice.AmountMb
+	deflateOnOom := req.BalloonDevice.DeflateOnOom
+	statsPollingIntervals := req.BalloonDevice.StatsPollingIntervals
+
+	balloonDevice := firecracker.NewBalloonDevice(amountMB, deflateOnOom, firecracker.WithStatsPollingIntervals(statsPollingIntervals))
+	return balloonDevice
+}
+
+func (s *service) getBalloonConfig(requestCtx context.Context, request *proto.GetVMInfoRequest) (*proto.FirecrackerBalloonDevice, error) {
+	b, err := s.machine.GetBalloonConfig(requestCtx)
+	balloonConfig := &proto.FirecrackerBalloonDevice{
+		AmountMb:              *b.AmountMb,
+		DeflateOnOom:          *b.DeflateOnOom,
+		StatsPollingIntervals: b.StatsPollingIntervals,
+	}
+	return balloonConfig, err
+}
+
+// UpdateBalloon will update an existing balloon device, before or after machine startup
+func (s *service) UpdateBalloon(requestCtx context.Context, req *proto.UpdateBalloonRequest) (*empty.Empty, error) {
+	defer logPanicAndDie(s.logger)
+
+	err := s.waitVMReady()
+	if err != nil {
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	s.logger.Info("updating balloon memory size")
+	if err := s.machine.UpdateBalloon(requestCtx, req.AmountMb); err != nil {
+		err = errors.Wrap(err, "failed to update memory balloon")
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	return &empty.Empty{}, nil
+}
+
+// UpdateBalloon will update an existing balloon device, before or after machine startup
+func (s *service) UpdateBalloonStats(requestCtx context.Context, req *proto.UpdateBalloonStatsRequest) (*empty.Empty, error) {
+	defer logPanicAndDie(s.logger)
+
+	err := s.waitVMReady()
+	if err != nil {
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	s.logger.Info("updating balloon statistics interval")
+	if err := s.machine.UpdateBalloonStats(requestCtx, req.StatsPollingIntervals); err != nil {
+		err = errors.Wrap(err, "failed to update balloon statistics interval")
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	return &empty.Empty{}, nil
+}
+
+// GetBalloonStats will return the latest balloon device statistics, only if enabled pre-boot.
+func (s *service) GetBalloonStats(requestCtx context.Context, req *proto.GetBalloonStatsRequest) (*proto.GetBalloonStatsResponse, error) {
+	defer logPanicAndDie(s.logger)
+
+	err := s.waitVMReady()
+	if err != nil {
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	s.logger.Info("Getting balloon statistics")
+	balloonStats, err := s.machine.GetBalloonStats(requestCtx)
+	if err != nil {
+		err = errors.Wrap(err, "failed to get balloon statistics")
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	// TODO: decide a better way for safely dereferencing?
+	if balloonStats.ActualMb == nil {
+		s.logger.Errorf("Failed to dereference nil ptr: alloonStats.ActualMb")
```

Comment:
Just a suggestion but instead of saying failed to dereference, maybe we can change the err message to saying that "X setting has not passed in the config".

---

Getting a bit more philosophical here, why do we want to catch these errs in the first place? Should we even let the init continue if these haven't been set by the client/ callers?

Reply:
Because these pointers should have some values associated with them if the client made a call, iirc